### PR TITLE
Copy keycloak files for oauth2 microservices

### LINKED ISSUE #8552: Gatling with OAuth2 and Keycloak
##### **Overview of the issue**

Gatling tests don't work with OAuth2 and Keycloak

##### **Motivation for or Use Case**

Here the stacktrace:

```
================================================================================
2018-10-13 10:06:39                                          90s elapsed
---- Requests ------------------------------------------------------------------
> Global                                                   (OK=19397  KO=0     )
> First unauthenticated request                            (OK=100    KO=0     )
> Authentication                                           (OK=100    KO=0     )
> Authentication Redirect 1                                (OK=100    KO=0     )
> patternfly.css Redirect 1                                (OK=100    KO=0     )
> favicon.ico Redirect 1                                   (OK=100    KO=0     )
> patternfly-additions.css Redirect 1                      (OK=100    KO=0     )
> zocial.css Redirect 1                                    (OK=100    KO=0     )
> login.css Redirect 1                                     (OK=100    KO=0     )
> Authenticate                                             (OK=100    KO=0     )
> Authenticate Redirect 1                                  (OK=100    KO=0     )
> Authenticate Redirect 2                                  (OK=100    KO=0     )
> Authenticate Redirect 3                                  (OK=100    KO=0     )
> Authenticate Redirect 4                                  (OK=100    KO=0     )
> Authenticate Redirect 5                                  (OK=100    KO=0     )
> Authenticate Redirect 6                                  (OK=100    KO=0     )
> Authenticate Redirect 7                                  (OK=100    KO=0     )
> Authenticate Redirect 8                                  (OK=100    KO=0     )
> Authenticate Redirect 9                                  (OK=100    KO=0     )
> Authenticate Redirect 10                                 (OK=100    KO=0     )
> Authenticate Redirect 11                                 (OK=100    KO=0     )
> Authenticate Redirect 12                                 (OK=100    KO=0     )
> Authenticate Redirect 13                                 (OK=100    KO=0     )
> Authenticate Redirect 14                                 (OK=100    KO=0     )
> Authenticate Redirect 15                                 (OK=100    KO=0     )
```

##### **Reproduce the error**

I generated the project with JHipster Online:

```
git clone https://github.com/pascalgrimaud/jhipster-test-oauth2
cd jhipster-test-oauth2
npm install
docker-compose -f src/main/docker/keycloak.yml up -d
./mvnw
```

In another Terminal:

```
gatling -sf src/test/gatling/user-files/simulations
```

##### **JHipster Version(s)**

5.4.2 and current master

##### **JHipster Version(s)**

```
jhipster@0.0.0 /home/pgrimaud/tmp/jhipster-test-oauth2
└── generator-jhipster@5.4.2 

```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "applicationType": "monolith",
    "gitCompany": "",
    "baseName": "jhipster",
    "packageName": "io.github.jhipster.application",
    "packageFolder": "io/github/jhipster/application",
    "serverPort": 8080,
    "serviceDiscoveryType": false,
    "authenticationType": "oauth2",
    "uaaBaseName": "../uaa",
    "cacheProvider": "ehcache",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "h2Disk",
    "prodDatabaseType": "mysql",
    "searchEngine": false,
    "enableSwaggerCodegen": false,
    "messageBroker": false,
    "buildTool": "maven",
    "useSass": false,
    "clientPackageManager": "npm",
    "testFrameworks": [
      "gatling",
      "protractor"
    ],
    "enableTranslation": false,
    "nativeLanguage": "en",
    "languages": [
      "en"
    ],
    "clientFramework": "angularX",
    "jhiPrefix": "jhi",
    "jhipsterVersion": "5.4.2"
  },
  "git-provider": "GitHub",
  "git-company": "pascalgrimaud",
  "repository-name": "jhipster-test-oauth2"
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>
entity Toto {
  name String
}

</pre>
</details>

##### **Environment and Tools**

openjdk version "1.8.0_181"
OpenJDK Runtime Environment (build 1.8.0_181-8u181-b13-0ubuntu0.18.04.1-b13)
OpenJDK 64-Bit Server VM (build 25.181-b13, mixed mode)

git version 2.19.1

node: v8.12.0

npm: 6.4.1

yeoman: 2.0.5

yarn: 1.10.1

Docker version 18.03.0-ce, build 0520e24

docker-compose version 1.21.2, build a133471
-   [x] Checking this box is mandatory (this is just to show you read everything)

### LINKED ISSUE #8553: Keycloak files not copied for oauth2 microservices
##### **Overview of the issue**

When creating a microservice with oauth2 option, the keycloak docker file and realms are not copied.

##### **Motivation for or Use Case**

When creating a microservice with oauth, I'd like to be able to run the entire service environment with docker-compose -f src/main/docker/app.yml

##### **Related issues**

n/a

##### **Suggest a Fix**

See [PR 8554](https://github.com/jhipster/generator-jhipster/pull/8554)

##### **JHipster Version(s)**

5.4.2 and master
 
##### **JHipster configuration**

```
jhipster@0.0.0 /home/rhermanns/code/jhipster/service
└── generator-jhipster@5.4.2 
```

##### **JHipster configuration, a `.yo-rc.json` file generated in the root folder**

<details>
<summary>.yo-rc.json file</summary>
<pre>
{
  "generator-jhipster": {
    "promptValues": {
      "packageName": "com.example.service",
      "nativeLanguage": "en"
    },
    "jhipsterVersion": "5.4.2",
    "applicationType": "microservice",
    "baseName": "service",
    "packageName": "com.example.service",
    "packageFolder": "com/example/service",
    "serverPort": "8081",
    "authenticationType": "oauth2",
    "cacheProvider": "hazelcast",
    "enableHibernateCache": true,
    "websocket": false,
    "databaseType": "sql",
    "devDatabaseType": "mariadb",
    "prodDatabaseType": "mariadb",
    "searchEngine": false,
    "messageBroker": false,
    "serviceDiscoveryType": "eureka",
    "buildTool": "gradle",
    "enableSwaggerCodegen": false,
    "jwtSecretKey": "bXktc2VjcmV0LXRva3VuLXRvLWNoYW5nZS1pbi1wcm9kdVN0aW9uLWFuZC10by1rZWVwLWluLWEtc2VjdXJlLXBsYWNl",
    "enableTranslation": true,
    "languages": [
      "en",
      "de"
    ],
    "testFrameworks": [
      "gatling"
    ],
    "jhiPrefix": "jhi",
    "clientPackageManager": "npm",
    "nativeLanguage": "en",
    "skipClient": true,
    "skipUserManagement": true
  }
}
</pre>
</details>

##### **JDL for the Entity configuration(s) `entityName.json` files generated in the `.jhipster` directory**

<details>
<summary>JDL entity definitions</summary>

<pre>

</pre>
</details>

##### **Environment and Tools**

java version "1.8.0_181"
Java(TM) SE Runtime Environment (build 1.8.0_181-b13)
Java HotSpot(TM) 64-Bit Server VM (build 25.181-b13, mixed mode)

git version 2.19.1

node: v10.12.0

npm: 6.4.1

Docker version 18.06.1-ce, build e68fc7a

docker-compose version 1.22.0, build f46880f

##### **Browsers and Operating System**

OSX

-   [X] Checking this box is mandatory (this is just to show you read everything)

### PR DESCRIPTION
**Pull request by rainerh**
- add missing realm-config files
- update copy conditions
- add tests for files existing
- remove previous cleanup task

Fixes #8553

-   Please make sure the below checklist is followed for Pull Requests.

-   [ ] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [X] Tests are added where necessary
-   [X] Documentation is added/updated where necessary
-   [X] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

### CODE REVIEW

File: generators/cleanup.js [hunk 1]
```diff
@@ -159,10 +159,5 @@ function cleanupOldServerFiles(generator, javaDir, testDir, mainResourceDir, tes
     if (generator.isJhipsterVersionLessThan('5.2.2')) {
         generator.removeFile(`${javaDir}config/ElasticsearchConfiguration.java`);
         generator.removeFile('gradle/liquibase.gradle');
-
-        if (generator.authenticationType === 'oauth2' && generator.applicationType === 'microservice') {
```
Comment (sendilkumarn) [inline generators/cleanup.js]: @rainerh It was removed on purpose in the previous commit - check this [conversation](https://github.com/jhipster/generator-jhipster/pull/8176#issuecomment-417708741). 
@mraible / @vishal423 can you please take a look at this?

Comment (rainerh) [reply generators/cleanup.js]: @sendilkumarn Have seen that, but with the current setup the keycloak.yml is referenced in the generated app.yml for microservices with oauth2.

Personally, I'd prefer to be able to run the entire stack for a microservice under development with all requirements, therefore I add most of the times my gateway as well. So I've the need for keycloak with its realm config.
Even if I just need to access the jhipster-registry, which of course has a UI, I'd need keycloak up and running.

If keycloak should not be added at all for microservices, then a other fix is required:

`
diff --git i/generators/server/templates/src/main/docker/app.yml.ejs w/generators/server/templates/src/main/docker/app.yml.ejs
index 508d1e889..26e70ed92 100644
--- i/generators/server/templates/src/main/docker/app.yml.ejs
+++ w/generators/server/templates/src/main/docker/app.yml.ejs
@@ -206,7 +206,7 @@ services:
         # Also set SPRING_CLOUD_CONSUL_CONFIG_FORMAT=files on your apps
         #    - CONFIG_MODE=git
     <%_ } _%>
-    <%_ if (authenticationType === 'oauth2') { _%>
+    <%_ if (authenticationType === 'oauth2' && applicationType !== 'microservice') { _%>
     keycloak:
         extends:
             file: keycloak.yml
`

**Comment by mraible** [comment]
We purposefully did not include Keycloak files in microservices. We expect you to only run Keycloak from your gateway.

**Comment by jdubois** [comment]
Yes I agree with @mraible here, having the same files copied everywhere should not be necessary and could cause many questions/issues

**Comment by rainerh** [comment]
So, if I have to run gateway with keycloak and registry from another repo in any case, to make proper use of the microservice setup, then the current setup would result in clashing eureka instances as it is configured right now.
Maybe I do not have full context, but having everything required available and preconfigured to start from the service under development perspective would help to reduce the complexity for service developers to think about starting gateway, keycloak and other dependecies from another repo. Not all developers have everything checked out in our teams, or even might even be unaware of this requirement, which causes also additional questions/issues.

On the other hand, if the gateway with keycloak is required to run from another repo, then the registry will be required as a dependency as well, so having registry here wouldn’t make that much sense anymore for me.

If keycloak files should not be copied - which is ok for me, then there still is the bug shown in the patch in my previous comment.

**Comment by jdubois** [comment]
Indeed, that's a very good point. We do provide the Registry files for microservices, so if we're logical we should indeed provide also the Keycloak files - so I'm changing my mind here.

**Comment by mraible** [comment]
The problem we experienced before was that the Keycloak files for the microservices have a different server port specified for the redirect URI since it uses the one for the microservice. So if you start the Keycloak instance in your microservice, then try to log in to it from your Gateway, you’ll get an invalid redirect URI error.

**Comment by rainerh** [comment]
I could also update the PR, so that no "infrastructure" docker yml and additional config files are copied over, and that developers have to start the required infrastructure via the gateway.

In case of a microservice, we should add some lines to the readme to describe how to start the required services, copy the simplified app.yml with the microservice under development and its persistence layer, maybe other, non general dependencies.
Would that be helpful?

**Comment by rainerh** [comment]
With #8675 applied, the above mentioned bug in my comment will be resolved.

Regarding the developer setup, the preference seems to be, that a gateway is required to work on a microservice.

In that case, shouldn't the file `src/main/docker/jhipster-registry.yml` be removed from miroservices?

If a gateway is required in any case, shouldn't it provide and start all required "global" infrastructure?
In my understanding these are:

- gateway-app
- gateway-db
- jhipster-registry
- keycloak (only in case of OAuth2)

The current gateway/app.yml, would possibly create port conflicts, because gateway-db is exposing its ports. 

@mraible, @jdubois Shall I update this PR in that direction?

**Comment by mraible** [comment]
Yes, I believe a gateway should be required to work on a microservice. However, I’m not sure everyone agrees.

**Comment by pascalgrimaud** [comment]
Closing this, as https://github.com/jhipster/generator-jhipster/pull/8675 has been merged instead

**Pull request closed without merging**